ForceAtlas2: continue #5213 for `prevent_overlapping` and `mobility`

eriknw opened the pull request:
This PR is a continuation of #5213 by @MathisHammel and aims to address issues that were identified in that PR and pay off some technical debt.

Some changes:
- Add support for passing vertex ids for initial positions, vertex_radius, and mobility.
- Add support for passing random_state.
- Add support for int64_t vertex/edge types.
- Remove support for `callback=` argument in `cugraph.force_atlas2` (it now gives informative error).
- Remove `cugraph/layout/force_atlas2_wrapper.pyx` (use PLC instead).
- Fix mutating input arrays (we now copy and use `const` in C++).
- Fix inputs and result when integral vertex ids weren't contiguous.
  - For example, if vertices are `[0, 2]`, inputs and result would contain values for `1`.
- Fixing renumbering of vertices.

This is getting close, only a couple failing tests to investigate, and I should add a new test or two.

MathisHammel committed:
ForceAtlas2 prevent_overlapping initial implementation

MathisHammel committed:
Fixed error message in force_atlas2_wrapper.pyx

MathisHammel committed:
Update force_atlas2.py to support multi-column vertex ids

MathisHammel committed:
Fixed wrong column name in force_atlas2_wrapper.pyx

MathisHammel committed:
Added mobility support to ForceAtlas2

MathisHammel committed:
Enabled prevent_overlapping with BH in ForceAtlas2

MathisHammel committed:
Removed mobility from speed computation functions

MathisHammel committed:
Added tests for ForceAtlas2 prevent_overlapping and mobility

MathisHammel committed:
Merge branch 'branch-25.10' of https://github.com/rapidsai/cugraph into fa2_overlap

MathisHammel committed:
Fixed style

MathisHammel committed:
Fixed Python style in ForceAtlas2

nv-rliu committed:
Merge branch 'branch-25.10' into fa2_overlap

MathisHammel committed:
Fixed multi-column indexing in FA2 mobility and noverlap

MathisHammel committed:
Removed dining_prefs from FA2 tests

eriknw committed:
Merge branch 'branch-25.10' into fa2_overlap

eriknw committed:
ForceAtlas2: continue #5213 for `prevent_overlapping` and `mobility`

This PR is a continuation of #5213 by @MathisHammel and aims to address issues
that were identified in that PR and pay off some technical debt.

**This PR is a work in progress**

Some changes:
- Add support for passing vertex ids for initial positions, vertex_radius, and mobility.
- Add support for passing random_state.
- Add support int64_t vertex/edge types.
- Remove support for `callback=` argument in `cugraph.force_atlas2` (it now gives informative error).
- Remove `cugraph/layout/force_atlas2_wrapper.pyx` (use PLC instead).
- Fix mutating input arrays (we now copy and use `const` in C++).
- Fix inputs and result when integral vertex ids weren't contiguous.
  - For example, if vertices are `[0, 2]`, inputs and result would contain values for `1`.
- Fixing renumbering of vertices.

Many things are still experimental and broken, but I am continuing to experiment and fix things :)

eriknw committed:
Merge branch 'branch-25.10' into fa2_overlap-contd

ChuckHastings commented on cpp/include/cugraph_c/layout_algorithms.h:
```diff
@@ -47,21 +47,28 @@ typedef struct {
  *                               Makes clusters more tight.
  * @param [in]   prevent_overlapping
  *                               Prevent nodes to overlap.
+ * @param [in]   vertex_radius_vertices
```
Do we want to make these separate from `start_vertices`?

Is there a use case where you specify x/y values and don't specify a vertex radius value, or vice verse?  Same question for mobility values.

I guess I'm less concerned with things that are O(number_of_vertices), but it does cost us GPU memory, and since they're managed by python the memory will be used for the duration of the call.

ChuckHastings commented on cpp/src/layout/legacy/barnes_hut.cuh:
```diff
@@ -93,17 +96,19 @@ void barnes_hut(raft::handle_t const& handle,
   rmm::device_uvector<float> d_minxl(blocks * FACTOR1, stream_view);
   rmm::device_uvector<float> d_minyl(blocks * FACTOR1, stream_view);
 
   // Actual mallocs
```
This old comment is not correct anymore.  The allocation happens in the RAII container construction above.  I'd suggest just deleting the comment.

eriknw replied:
Done.

ChuckHastings commented on cpp/src/layout/legacy/barnes_hut.cuh:
```diff
@@ -162,9 +165,19 @@ void barnes_hut(raft::handle_t const& handle,
+  graph.degree(massl_et, cugraph::legacy::DegreeDirection::OUT);
   RAFT_CHECK_CUDA(stream_view.value());
 
+  // Cast massl from edge_t to int.
```
We should probably look at updating the functions to use `edge_t`.  While it is uncommon, it is possible for an edge in a 64-bit graph to have > 2^31-1 neighbors, in which case we'll generate an incorrect answer by down casting this into a 32-bit integer.  Not sure the ripple effects here... my first glance is that it seems tractable.

eriknw replied:
Thanks, working on this now! It's straightforward so far

ChuckHastings commented on cpp/src/layout/legacy/exact_fa2.cuh:
```diff
@@ -96,9 +100,19 @@ void exact_fa2(raft::handle_t const& handle,
   RAFT_CHECK_CUDA(stream_view.value());
 
+  // Cast mass from edge_t to int.
+  // ERIK: this is a bandaid since `graph.degree` uses edge_t and other
```
Same

ChuckHastings commented on cpp/src/layout/legacy/fa2_kernels.cuh:
```diff
@@ -83,16 +106,18 @@ void apply_attraction(const vertex_t* restrict row,
 
   dim3 nthreads, nblocks;
-  nthreads.x = min(e, CUDA_MAX_KERNEL_THREADS);
+  nthreads.x = min(static_cast<int>(e), CUDA_MAX_KERNEL_THREADS);
```
Would this be better as:
`nthreads.x = static_cast<int>(e, CUDA_MAX_KERNEL_THREADS)`?

If e doesn't fit in in a 32-bit integer we'll want to be sure we pick CUDA_MAX_KERNEL_THREADS.

If the above doesn't work, something like this might work:

`nthreads.x = static_cast<int>(e, static_cast<edge_t>(CUDA_MAX_KERNEL_THREADS))`?

ChuckHastings commented on cpp/src/layout/legacy/fa2_kernels.cuh:
```diff
@@ -83,16 +106,18 @@ void apply_attraction(const vertex_t* restrict row,
   nthreads.y = 1;
   nthreads.z = 1;
-  nblocks.x  = min((e + nthreads.x - 1) / nthreads.x, CUDA_MAX_BLOCKS);
+  nblocks.x  = min((static_cast<int>(e) + nthreads.x - 1) / nthreads.x, CUDA_MAX_BLOCKS);
```
Same concern

ChuckHastings commented on cpp/src/layout/legacy/fa2_kernels.cuh:
```diff
@@ -168,10 +195,10 @@ void apply_gravity(const float* restrict x_pos,
 {
   dim3 nthreads, nblocks;
-  nthreads.x = min(n, CUDA_MAX_KERNEL_THREADS);
+  nthreads.x = min(static_cast<int>(n), CUDA_MAX_KERNEL_THREADS);
```
Same

ChuckHastings commented on cpp/src/layout/legacy/fa2_kernels.cuh:
```diff
@@ -168,10 +195,10 @@ void apply_gravity(const float* restrict x_pos,
   nthreads.y = 1;
   nthreads.z = 1;
-  nblocks.x  = min((n + nthreads.x - 1) / nthreads.x, CUDA_MAX_BLOCKS);
+  nblocks.x  = min((static_cast<int>(n) + nthreads.x - 1) / nthreads.x, CUDA_MAX_BLOCKS);
```
Same

ChuckHastings commented on cpp/src/layout/legacy/fa2_kernels.cuh:
```diff
@@ -222,10 +249,10 @@ void compute_local_speed(const float* restrict repel_x,
 {
   dim3 nthreads, nblocks;
-  nthreads.x = min(n, CUDA_MAX_KERNEL_THREADS);
+  nthreads.x = min(static_cast<int>(n), CUDA_MAX_KERNEL_THREADS);
```
Same

ChuckHastings commented on cpp/src/layout/legacy/fa2_kernels.cuh:
```diff
@@ -222,10 +249,10 @@ void compute_local_speed(const float* restrict repel_x,
   nthreads.y = 1;
   nthreads.z = 1;
-  nblocks.x  = min((n + nthreads.x - 1) / nthreads.x, CUDA_MAX_BLOCKS);
+  nblocks.x  = min((static_cast<int>(n) + nthreads.x - 1) / nthreads.x, CUDA_MAX_BLOCKS);
```
Same

ChuckHastings commented on cpp/src/layout/legacy/fa2_kernels.cuh:
```diff
@@ -306,20 +335,31 @@ void apply_forces(float* restrict x_pos,
 {
   dim3 nthreads, nblocks;
-  nthreads.x = min(n, CUDA_MAX_KERNEL_THREADS);
+  nthreads.x = min(static_cast<int>(n), CUDA_MAX_KERNEL_THREADS);
```
Same

ChuckHastings commented on cpp/src/layout/legacy/fa2_kernels.cuh:
```diff
@@ -306,20 +335,31 @@ void apply_forces(float* restrict x_pos,
   nthreads.y = 1;
   nthreads.z = 1;
-  nblocks.x  = min((n + nthreads.x - 1) / nthreads.x, CUDA_MAX_BLOCKS);
+  nblocks.x  = min((static_cast<int>(n) + nthreads.x - 1) / nthreads.x, CUDA_MAX_BLOCKS);
```
Same

eriknw committed:
Respond to feedback: propagate mass type, etc.

Also, check types of inputs in `cugraph.force_atlas` and convert/warn as appropriate.

eriknw committed:
Make `srcs` and `dsts` in `c_api/legacy_fa2.cpp` the renumbered indices

eriknw committed:
Merge branch 'branch-25.10' into fa2_overlap-contd

eriknw committed:
Fix tests and remove force_atlas2.pxd from cugraph

eriknw committed:
Re-add dining_prefs to FA2 tests (revert cc013ca)

To fix tests, had to change usage of `G.nodes()`

eriknw committed:
Merge branch 'branch-25.10' into fa2_overlap-contd

eriknw commented:
This PR is ready for review. I left a few questions inline.

I plan to update nx-cugraph next, and maybe add another test or two here. Are there any other changes we should make while I'm familiar with the code for FA2? Some possibilities:
- Re-add support for the `callback` parameter (by adding it to PLC)
- Allow `vertex_radius_vertices` (as used by PLC) to be a subset of vertices; unspecified can default to 0
- Allow `mobility_vertices` (as used by PLC) to be a subset of vertices; unspecified can default to 1
- Allow `start_vertices` (as used by PLC) to be a subset of vertices; unsure what to do for unspecified--random?

ChuckHastings commented on python/cugraph/cugraph/layout/force_atlas2.py:
```diff
@@ -129,43 +187,109 @@ def on_train_end(self, positions):
+        if input_graph.renumbered:
+            pos_list = renumber_vertices(input_graph, pos_list, 2)
+        # Ensure dtypes are valid, warn if we need to cast
+        initial_pos_vertices = ensure_vertex_dtype(
+            input_graph, pos_list["vertex"], 'pos_list["vertex"]'
+        )
+        initial_pos_x = ensure_float32_dtype(pos_list["x"], 'pos_list["x"]')
+        initial_pos_y = ensure_float32_dtype(pos_list["y"], 'pos_list["y"]')
```
Yes, I notice a bug in the original C API code where it checks to make sure that the types of xstart and ystart match, but then assumes that they are of type `float`.

I would modify the CAPI_EXPECTS call to verify that all of the parameters are of type `float`.'

FYI - edge dtype is either int32_t or int64_t and must match vertex dtype.  The weight would be the only floating point type in the legacy graph structure.  Does FA2 use weights?  If so, it would probably be necessary to make sure that weight_t is also `float`... although if that were the case I would expect you would have hit a compile error.

eriknw replied:
Aha, yes indeed I'm "creating a graph with integral edge types from python and the results are reasonable", and yes indeed the weights are converted to floating point.

ChuckHastings commented on cpp/src/c_api/legacy_fa2.cpp:
```diff
@@ -167,48 +219,123 @@ struct force_atlas2_functor : public cugraph::c_api::abstract_functor {
+          y_start.data(), y_start_->as_type<float>(), y_start_->size_, handle_.get_stream());
 
-        number_map_pos = rmm::device_uvector<vertex_t>{number_map->size(), handle_.get_stream()};
+        cugraph::renumber_ext_vertices<vertex_t, multi_gpu>(
```
SG-only shouldn't matter, I tend to use `renumber_ext_vertices` in SG code, because there's really no concept of a local vertex in SG.

In MG the `renumber_local_ext_vertices` is more efficient because you are asserting that all of the vertices you are renumbering are local to this GPU and you don't have to do any shuffling of data.  But sometimes the results are partitioned on the GPU where you can't guarantee that, so if you call `renumber_ext_vertices` it will shuffle the necessary data to the right GPU to get the renumbering done properly.

eriknw replied:
Great, thanks for explaining 👍

ChuckHastings reviewed (commented):
I'll approve C++ changes after you remove the alternative options that I commented on.

rlratzel reviewed (commented):
Thanks, Erik! I'm excited to have less tech debt and the legacy Python API removed. I had a few suggestions/questions.

eriknw commented on python/cugraph/cugraph/layout/force_atlas2.py:
```diff
@@ -129,43 +187,109 @@ def on_train_end(self, positions):
             raise TypeError("pos_list should be a cudf.DataFrame")
-        if set(pos_list.columns) != set(["x", "y", "vertex"]):
+
+        if set(pos_list.columns) != {"x", "y", "vertex"}:
```
That's a good idea! I actually started changing this early on, but stopped b/c other things needed more attention. I'll change it so a single dataframe can be used for all.

eriknw replied:
Done.

eriknw committed:
Add more `CAPI_EXPECTS` and let same input df be used

eriknw committed:
Merge branch 'branch-25.10' into fa2_overlap-contd

eriknw commented:
/ok to test

eriknw commented on python/cugraph/cugraph/layout/force_atlas2.py:
```diff
@@ -23,15 +64,20 @@ def force_atlas2(
     outbound_attraction_distribution=True,
     lin_log_mode=False,
     prevent_overlapping=False,
+    vertex_radius=None,
+    overlap_scaling_ratio=100.0,
     edge_weight_influence=1.0,
     jitter_tolerance=1.0,
     barnes_hut_optimize=True,
     barnes_hut_theta=0.5,
     scaling_ratio=2.0,
     strong_gravity_mode=False,
     gravity=1.0,
+    mobility=None,
```
Why are the new parameters `vertex_radius` and `mobility`? Would e.g. `vertex_mobility` be more clear and consistent?

rlratzel replied:
> Why are the new parameters `vertex_radius` and `mobility`? Would e.g. `vertex_mobility` be more clear and consistent?

Sounds good to me!

rlratzel replied:
@eriknw shall I `/merge` or do you want to make this change?

rlratzel replied:
I agree that it seems more consistent so I'm in favor.
I'll add a `DO_NOT_MERGE` since this has been approved so it doesn't get merged by someone else before you get that change in.

eriknw replied:
I pushed the change and removed the `DO_NOT_MERGE` label

eriknw committed:
Merge branch 'branch-25.10' into fa2_overlap-contd

nv-rliu commented:
/ok to test

rlratzel commented:
Added `DO_NOT_MERGE` since there's one more change to be made: https://github.com/rapidsai/cugraph/pull/5260#discussion_r2380624038

eriknw committed:
Rename `mobility` to `vertex_mobility`

eriknw commented:
/ok to test [8a12d44](https://github.com/rapidsai/cugraph/pull/5260/commits/8a12d440bbede266df5a85468a65cb2fd1bd8f41)

eriknw commented on python/cugraph/cugraph/layout/force_atlas2.py:
```diff
@@ -129,43 +187,117 @@ def on_train_end(self, positions):
+            vertex_mobility["mobility"], 'vertex_mobility["mobility"]'
+        )
 
     if input_graph.is_directed():
         input_graph = input_graph.to_undirected()
```
Would it be better to symmetrize the edges in C++?
